cassiopeia/solver/dissimilarity_functions.py

```diff
@@ -70,3 +69,49 @@ def weighted_hamming_distance(
         return 0
 
     return d / num_present
+
+def hamming_similarity(
+    s1: List[int], 
+    s2: List[int],
+    missing_char: int,
+    w: Optional[Dict[int, Dict[int, float]]] = None,
+) -> float:
+    """A function to return the number of (non-missing) character/state
+    mutations shared by two samples.
+
+    Args:
+        s1: Character states of the first sample
+        s2: Character states of the second sample
+        missing_char: The character representing missing values
+        w: A set of optional weights to weight the similarity of a mutation
+    Returns:
+        The number of shared mutations between two samples, weighted or unweighted
+    """
+
+    # TODO Optimize this using masks
+    num_chars = len(s1)
+    assert len(s2) == num_chars
+    if w is None:
+        return sum(
+            [
+                1
+                for i in range(num_chars)
+                if s1[i] == s2[i]
+                and (
+                    s1[i] != 0
+                    and s1[i] != missing_char
+                )
+            ]
+        )
+    else:
+        return sum(
+            [
+                w[i][s1[i]]
+                for i in range(num_chars)
+                if s1[i] == s2[i]
+                and (
+                    s1[i] != 0
+                    and s1[i] != missing_char
+                )
+            ]
+        )
```

Comment:
I think this is going to be redundant with my `hamming_distance` function. Why not use the weighted hamming dissimilarity function that I wrote previously?

Reply:
It's slightly different from your weighted hamming dissimilarity in that it only increments the score +1 when you have a match, when in the case of the weighted hamming it increments +2. And it doesn't have the case where if the characters are different but one of them is 0, then it increments +1. I think that this will likely not be the best similarity to use for spectral in the future, but I kept it for now. When we get benchmarks that show that the weighted hamming distance allows for better performance, I think then I'll remove it

Reply:
It's a more naive notion of similarity for sure

Reply:
Okay, let's keep it for now but can you change the name to `hamming_distance_without_missing` or something so a user is not confused?